Crypto IPC 64 bit key ids for ITS

itayzafrir opened the pull request:
### Description
Support 64 bit crypto key ids for TARGET_PSA systems.

Crypto type `psa_key_id_t` is defined differently between client and server, this PR is here to provide the key owner (i.e. the calling partition) for use by the crypto library.

This PR assures that a crypto key can be opened only by the key owner/creator, i.e. the partition which created the key in the first place.

~~**Please do not merge or run CI** - depends on https://github.com/ARMmbed/mbed-crypto/pull/32 and also https://github.com/ARMmbed/mbed-crypto/pull/59. 
Once 32 and 59 have been merged to mbed-crypto this PR will need to be re-imported via crypto importer into Mbed OS.~~

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@avolinski @Patater 
<!-- 
    Optional
    Request additional reviewers with @username
-->

Commit:
crypto service: Remove unused IPC argument and unneeded casts

Commit:
crypto service: Assemble bit crypto key ids

ciarmcom commented:
@itayzafrir, thank you for your changes.
@Patater @avolinski @ARMmbed/mbed-os-core @ARMmbed/mbed-os-tools @ARMmbed/mbed-os-maintainers @ARMmbed/mbed-os-test @ARMmbed/mbed-os-crypto @ARMmbed/mbed-os-tls please review.

itayzafrir commented:
Rebased on master, please take a look guys.

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_PSA_SRV_IPC/psa_crypto_spm.c:
```diff
@@ -461,14 +461,14 @@ psa_status_t psa_aead_encrypt(psa_key_handle_t key_handle,
+        status = PSA_ERROR_COMMUNICATION_FAILURE;
     }
-    return ((psa_status_t)call_error);
+    return ((psa_status_t)status);
```
no cast needed :)

itayzafrir replied:
fixed

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_PSA_SRV_IPC/psa_crypto_spm.c:
```diff
@@ -533,14 +533,14 @@ psa_status_t psa_aead_decrypt(psa_key_handle_t key_handle,
+        status = PSA_ERROR_COMMUNICATION_FAILURE;
     }
-    return ((psa_status_t)call_error);
+    return ((psa_status_t)status);
```
no cast needed :)

itayzafrir replied:
fixed

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_PSA_SRV_IPC/psa_crypto_spm.c:
```diff
@@ -584,14 +584,14 @@ psa_status_t psa_asymmetric_sign(psa_key_handle_t key_handle,
+        status = PSA_ERROR_COMMUNICATION_FAILURE;
     }
-    return ((psa_status_t)call_error);
+    return ((psa_status_t)status);
```
no cast needed :)

itayzafrir replied:
fixed

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_PSA_SRV_IPC/psa_crypto_spm.c:
```diff
@@ -624,13 +624,13 @@ psa_status_t psa_asymmetric_verify(psa_key_handle_t key_handle,
+        status = PSA_ERROR_COMMUNICATION_FAILURE;
     }
-    return ((psa_status_t)call_error);
+    return ((psa_status_t)status);
```
no cast needed :)

itayzafrir replied:
fixed

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_PSA_SRV_IPC/psa_crypto_spm.c:
```diff
@@ -684,14 +684,14 @@ static psa_status_t psa_asymmetric_operation(psa_sec_function_t func,
+        status = PSA_ERROR_COMMUNICATION_FAILURE;
     }
-    return ((psa_status_t)call_error);
+    return ((psa_status_t)status);
```
no cast needed :)

itayzafrir replied:
fixed

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_PSA_SRV_IPC/psa_crypto_spm.c:
```diff
@@ -1470,11 +1470,11 @@ psa_status_t psa_cipher_set_iv(psa_cipher_operation_t *operation,
+        status = PSA_ERROR_COMMUNICATION_FAILURE;
     }
-    return ((psa_status_t) err);
+    return (status);;
```
```suggestion
    return (status);
```

itayzafrir replied:
fixed

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_PSA_SRV_IPC/psa_crypto_spm.c:
```diff
@@ -1442,18 +1442,18 @@ psa_status_t psa_cipher_generate_iv(psa_cipher_operation_t *operation,
+        status = PSA_ERROR_COMMUNICATION_FAILURE;
     }
-    return ((psa_status_t) err);
+    return (status);;
```
```suggestion
    return (status);
```

itayzafrir replied:
fixed

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_PSA_SRV_IPC/psa_crypto_spm.c:
```diff
@@ -1379,18 +1379,18 @@ psa_status_t psa_cipher_encrypt_setup(psa_cipher_operation_t *operation,
+        status = PSA_ERROR_COMMUNICATION_FAILURE;
     }
-    return ((psa_status_t) err);
+    return (status);;
```
```suggestion
    return (status);
```

itayzafrir replied:
fixed

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_SPE/crypto_types_spe.h:
```diff
@@ -0,0 +1,101 @@
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+#ifndef PSA_CRYPTO_TYPES_H
```
why can't we copy the original file in importer and apply git hash patch on it for our small change?

itayzafrir replied:
It's possible, but it would complicate things. Also, the importer script does not modify/copy files outside the crypto scope, these files reside in mbed-os/components/...
Maybe we could do what you suggested or automate in some other way it sometime later?

alekshex replied:
for next version we need to rethink it

itayzafrir commented:
@avolinski I've rebased and addressed your comments so far. Old branch @ https://github.com/itayzafrir/mbed-os/tree/crypto-64-bit-key-ids-1

NirSonnenschein commented:
@ARMmbed/mbed-os-core @ARMmbed/mbed-os-tls @ARMmbed/mbed-os-tools @ARMmbed/mbed-os-test @ARMmbed/mbed-os-crypto please review or comment if you don't wish to review.

bridadan reviewed (commented):
There are no tool changes in this PR

Commit:
Define MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER

This enables crypto encoding an owner in key file IDs.
Added a static assert check in client side proxy.

Commit:
Remove duplicate inclusion of header files

Remove duplicate inclusion of psa/client.h and psa/service.h

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_PSA_SRV_IPC/psa_crypto_spm.c:
```diff
@@ -23,31 +23,37 @@
 
+#include "mbed_assert.h"
+
 #define MINOR_VER 1
```
aren't we changing minor version with this commit? need to be careful here for bc

itayzafrir replied:
no need

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_SPE/crypto_spe.h:
```diff
@@ -65,6 +65,8 @@ extern "C" {
 #define psa_close_key  psa_sec_close_key
 #define psa_hash_clone  psa_sec_hash_clone
 
+#define MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER 1
```
doesn't seem to be used.
also can't understand what is encodes_owner

itayzafrir replied:
it's used for service integration with the crypto library as of https://github.com/ARMmbed/mbed-crypto/pull/59

alekshex commented on components/TARGET_PSA/services/crypto/COMPONENT_SPE/psa_crypto_partition.c:
```diff
@@ -1172,16 +1179,16 @@ static void psa_key_management_operation(void)
-                    bytes_read = psa_read(msg.handle, 2, &max_bits, msg.in_size[2]);
-                    if (bytes_read != msg.in_size[2]) {
-                        SPM_PANIC("SPM read length mismatch");
+
```
previously there was:
bytes_read = psa_read(msg.handle, 2, &max_bits, msg.in_size[2]);
what was passed in msg.in_size[2]?

itayzafrir replied:
`max_bits` is no longer part of the crypto API, so it was removed

Patater commented on components/TARGET_PSA/services/crypto/COMPONENT_SPE/psa_crypto_partition.c:
```diff
@@ -1213,16 +1220,16 @@ static void psa_key_management_operation(void)
+                    id.owner = psa_identity(msg.handle);
 
-                    bytes_read = psa_read(msg.handle, 1, &id, msg.in_size[1]);
+                    bytes_read = psa_read(msg.handle, 1, &(id.key_id), msg.in_size[1]);
```
Consider using `PSA_KEY_FILE_GET_KEY_ID` to get the key ID.

itayzafrir replied:
That wouldn't work here. `PSA_KEY_FILE_GET_KEY_ID` returns the value of `key_id`, here we're reading from the incoming message into `key_id`.

Patater replied:
`&PSA_KEY_FILE_GET_KEY_ID(id)` wouldn't work?

The downside to how this is written now is it doesn't get the advantage of the macro abstraction and may be depending on more implementation detail than it needs to. This isn't a super strong downside now, but you should be aware of and consider it.

itayzafrir replied:
My bad, it would work. However I'd like to leave it as is for now, since there's no abstraction for the `owner`.

Patater replied:
OK

Patater reviewed (approved):
OK
